### ISSUE TITLE
Accuracy_Calculator on large datasets.

### ISSUE THREAD

**Issue by chichieinstein** [KevinMusgrave/pytorch-metric-learning#566]
Hello.

I am trying to use metric learning on a dataset that contains ~120K datapoints. Each datapoint is represented by features that are numpy arrays of dimension (3, 4096, 128) and each entry is a float32. Batchsize is 256 (I have access to 8GPUs each of capacity ~22GiB) and embedding size is 100. The number of classes is 6.

When I try to use Accuracy calculation at the end of each epoch, I run into CUDA memory error, if I try to use the faiss-gpu version of accuracy calculation, and if I try to use the faiss-cpu version, it seems the total number of triplets at the end of epoch do not fit into RAM, and the server kills the process. I have set `k` to max_bin_count.

Short of reducing the size of the dataset (I had successfully run this training for the same features when the dataset size was smaller) is there anything else smart I can do? Thank you in advance.

Just to check if my understanding is correct, if batch size is `B` and number of iterations is `I` and the embedding size is `S`, then the number of triplets should scale as `O(B^3I)` correct? And the memory usage of triplets should scale as `O(B^3 I S)`?

Finally, I am also experimenting with turning mining off and using SoftTripletLoss. If I set k=1000 in the accuracy calculation, say, that should mean only 1000 reference embeddings are loaded into memory right? That would of course mean that MAP@R will be inaccurate.

**Comment by KevinMusgrave** [comment]
> When I try to use Accuracy calculation at the end of each epoch, I run into CUDA memory error, if I try to use the faiss-gpu version of accuracy calculation, and if I try to use the faiss-cpu version, it seems the total number of triplets at the end of epoch do not fit into RAM, and the server kills the process. I have set `k` to max_bin_count.

> Short of reducing the size of the dataset (I had successfully run this training for the same features when the dataset size was smaller) is there anything else smart I can do? Thank you in advance.


If faiss is causing the out-of-memory issue, you could try CustomKNN, which allows computing k-nearest-neighbors in batches:

```python
from pytorch_metric_learning.distances import LpDistance
from pytorch_metric_learning.utils.inference import CustomKNN

distance_fn = LpDistance(normalize_embeddings=False, power=2)
custom_knn = CustomKNN(distance_fn, batch_size=32)
ac = AccuracyCalculator(knn_func=custom_knn)
```


> Just to check if my understanding is correct, if batch size is `B` and number of iterations is `I` and the embedding size is `S`, then the number of triplets should scale as `O(B^3I)` correct? And the memory usage of triplets should scale as `O(B^3 I S)`?

The AccuracyCalculator does not use triplets. Are you referring to memory usage during training?

> Finally, I am also experimenting with turning mining off and using SoftTripletLoss. If I set k=1000 in the accuracy calculation, say, that should mean only 1000 reference embeddings are loaded into memory right? That would of course mean that MAP@R will be inaccurate.

The miner and loss functions are only used during training. They have no effect on AccuracyCalculator.

The number of reference embeddings is the number of embeddings you pass into the `get_accuracy` function of AccuracyCalculator. Let's say you pass in 20,000 query embeddings and 100,000 reference embeddings. Setting `k=1000` means that for each of the 20,000 query embeddings, it will find the 1000 nearest reference embeddings from the pool of 100,000 reference embeddings.

**Comment by KevinMusgrave** [comment]
Glad to hear it's working now
